logins-sql/src/db.rs

```diff
@@ -0,0 +1,844 @@
+        let mut outgoing = OutgoingChangeset::new("passwords".into(), st);
+        let mut stmt = self.db.prepare_cached(&format!("
+            SELECT * FROM {local}
+            WHERE sync_status IS NOT {synced}",
+            local = schema::LOCAL_TABLE_NAME,
+            synced = SyncStatus::Synced as u8
+        ))?;
+        let rows = stmt.query_and_then(&[], |row| {
+            Ok(if row.get::<_, bool>("is_deleted") {
+                Payload::new_tombstone(row.get_checked::<_, String>("guid")?)
+            } else {
+                let login = Login::from_row(row)?;
+                Payload::from_record(login)?
+            })
+        })?;
+        outgoing.changes = rows.collect::<Result<_>>()?;
+
+        Ok(outgoing)
+    }
+
+    fn do_apply_incoming(
+        &mut self,
+        inbound: IncomingChangeset
+    ) -> Result<OutgoingChangeset> {
+        let data = self.fetch_login_data(&inbound.changes)?;
+        let plan = self.reconcile(data, inbound.timestamp)?;
+        self.execute_plan(plan)?;
+        Ok(self.fetch_outgoing(inbound.timestamp)?)
+    }
+
+    fn put_meta(&self, key: &str, value: &ToSql) -> Result<()> {
+        self.execute_cached_with_args(&*META_PUT_SQL, &[&key as &ToSql, value])?;
+        Ok(())
+    }
+
+    fn get_meta<T: FromSql>(&self, key: &str) -> Result<Option<T>> {
+        self.query_row_cached(&*META_GET_SQL, &[&key as &ToSql], |row| Ok(row.get(0)))
+    }
+
+    pub fn set_last_sync(&self, last_sync: ServerTimestamp) -> Result<()> {
+        debug!("Updating last sync to {}", last_sync);
+        self.put_meta(schema::LAST_SYNC_META_KEY, &last_sync.0)
+    }
+
+    pub fn set_global_state(&self, global_state: &str) -> Result<()> {
+        self.put_meta(schema::GLOBAL_STATE_META_KEY, &global_state)
+    }
+
+    pub fn get_last_sync(&self) -> Result<Option<ServerTimestamp>> {
+        Ok(self.get_meta::<f64>(schema::LAST_SYNC_META_KEY)?.map(ServerTimestamp))
+    }
+
+    pub fn get_global_state(&self) -> Result<Option<String>> {
+        self.get_meta::<String>(schema::GLOBAL_STATE_META_KEY)
+    }
+}
+
+impl Store for LoginDb {
+    type Error = Error;
+
+    fn apply_incoming(
+        &mut self,
+        inbound: IncomingChangeset
+    ) -> Result<OutgoingChangeset> {
+        self.do_apply_incoming(inbound)
+    }
+
+    fn sync_finished(
+        &mut self,
+        new_timestamp: ServerTimestamp,
+        records_synced: &[String],
+    ) -> Result<()> {
+        self.mark_as_synchronized(
+            &records_synced.iter().map(|r| r.as_str()).collect::<Vec<_>>(),
+            new_timestamp
+        )
+    }
+}
+
+lazy_static! {
+
+    static ref ANY_SYNCED_SQL: String = format!("
+        SELECT EXISTS(
+            SELECT 1 from {mirror}
+            UNION ALL
+            SELECT 1 from {local} WHERE sync_status IS NOT {new}
+        )",
+        mirror = schema::MIRROR_TABLE_NAME,
+        local = schema::LOCAL_TABLE_NAME,
+        new = SyncStatus::New as u8
```

Comment:
I have a gentle preference for dangling commas to remove two-line diffs in the future, but I don't feel strongly about it.